### PR TITLE
DX-294 Order Integration

### PR DESCRIPTION
**Pull request by seanbocz**
Added Magento URLs needed for Order Integration.
Added EA URLs needed for Orders.
Added Models required for Order Integration flow.
Added controllers for handling calls required for Order sync process. 
Added Filters that can be used for Hypermedia service requests.
Added Mappers needed for logic in order integration.
Separated logs into Order and Product logs. 
Adjusted time for sync logging so that in the event that a product / order is adjusted during a sync, it will be updated on the next sync.
Created unit tests for new controllers to ensure they all run and return the expected data.
Added tests for CustomerMapper, EntityMapper, and OrderMapper.
@danw-iq @melissakendall

### CODE REVIEW

File: Controllers/Magento/ProductController.cs [hunk 1]
```diff
@@ -161,7 +161,7 @@ public void AddCustomAttributeToProduct(ProductResource magentoProduct, List<int
 		/// Gets the catalog inventory item based on the SKU provided.
 		/// </summary>
 		/// <param name="sku">SKU associated with the catalog inventory item</param>
-		/// <returns>Gatalog inventory item based on the SKU provided</returns>
+		/// <returns>Catalog inventory item based on the SKU provided</returns>
 		public CatalogInventoryItemResource GetInventoryBySku(string sku)
```
Comment (melissakendall) [inline Controllers/Magento/ProductController.cs]: "inventory" can be ambiguous, lets call this Quantity in Magento context app-wide

File: Controllers/Magento/CustomerController.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+		public CustomerController(string magentoAuthToken)
+		{
+			MagentoAuthToken = magentoAuthToken;
+		}
+
+		/// <summary>
+		/// Gets the Magento Customer specified
+		/// </summary>
+		/// <param name="customerId">Customer to return</param>
```
Comment (melissakendall) [inline Controllers/Magento/CustomerController.cs]: false, this accepts a customer identifier, not a customer =p

File: Controllers/Magento/CartController.cs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+		public CartController(string magentoAuthToken)
+		{
+			MagentoAuthToken = magentoAuthToken;
+		}
+
+		/// <summary>
+		/// Creates a cart / quote for the customer specified in Magento and returns that cart's ID
+		/// </summary>
+		/// <param name="customerId">Customer to create cart for</param>
```
Comment (melissakendall) [inline Controllers/Magento/CartController.cs]: identifier of a ..

File: Controllers/Magento/CartController.cs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+			//Ensure we get the right code
+			CheckStatusCode(response.StatusCode);
+
+			return JsonConvert.DeserializeObject<int>(response.Content);
+		}
+
+		/// <summary>
+		/// Adds an item to a cart in Magento and returns the CartItemResource created
+		/// Throws exception if invalid CartAddItemResource is passed in
```
Comment (melissakendall) [inline Controllers/Magento/CartController.cs]: specify what an "invalid" would be, specifically what situations might cause this: if its an unlikely edge case or something that could happen

File: Controllers/Magento/CartController.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+		/// </summary>
+		/// <param name="cartId">Cart to add item to</param>
+		/// <param name="item">Item to add to cart</param>
+		/// <returns>Cart Item that was added</returns>
+		public CartItemResource AddItemToCart(int cartId, CartAddItemResource item)
+		{
+			if(item.cartItem.quote_id == null || string.IsNullOrEmpty(item.cartItem.sku))
+				throw new Exception("Ensure that the body of the CartAddItemResource used is properly initialized.");
```
Comment (melissakendall) [inline Controllers/Magento/CartController.cs]: this message could be more useful - how does the user know the cart is initialized? what needs to happen for this message to appear? its better to have messages that specify what the user needs to do to fix (try again, remove bad data, etc)

File: Controllers/EndlessAisle/EntitiesController.cs [hunk 6]
```diff
@@ -6,36 +6,57 @@
...
+	{
+		public static string EndlessAisleAuthToken;
+
+		public EntitiesController(string eaAuthToken)
+		{
+			EndlessAisleAuthToken = eaAuthToken;
+		}
+
+		/// <summary>
+		/// Gets information about all manufacturers
```
Comment (melissakendall) [inline Controllers/EndlessAisle/EntitiesController.cs]: I notice I made this original comment, but lets change it to "a list of supported Manufacturers in EA"

File: Controllers/EndlessAisle/EntitiesController.cs [hunk 7]
```diff
@@ -6,36 +6,57 @@
...
+			var response = client.Execute(request);
+
+			//Ensure we get the right code
+			CheckStatusCode(response.StatusCode);
+
+			return JsonConvert.DeserializeObject<List<ManufacturerResource>>(response.Content);
+		}
+
+		/// <summary>
+		/// Gets the EA location set in the App.config file
```
Comment (melissakendall) [inline Controllers/EndlessAisle/EntitiesController.cs]: to hammer it home, specify that this is the location the EA device is located at, also Returns comment is empty :p

File: Mappers/EntityMapper.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+		private readonly RegionController _magentoRegionController;
+
+		public EntityMapper(string magentoAuthToken, string eaAuthToken) : base(magentoAuthToken, eaAuthToken)
+		{
+			_eaEntitiesController = new EntitiesController(eaAuthToken);
+			_magentoRegionController = new RegionController(magentoAuthToken);
+		}
+
+		//Magento Region from the config file
+		public RegionResource MagentoRegion
```
Comment (melissakendall) [inline Mappers/EntityMapper.cs]: describe the algorithm this uses, although it is trivial to us, to an external dev it may not be

File: Mappers/OrderMapper.cs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		/// Returns a list of the EA orders that have been created since the specified time
+		/// </summary>
+		/// <param name="createdAfter">Get orders created after this time</param>
+		/// <returns>List of the EA orders that match the time criteria</returns>
+		public IEnumerable<OrderResource> GetEaOrdersCreatedAfter(DateTime createdAfter)
+		{
+			if (createdAfter > DateTime.UtcNow)
+				throw new ArgumentOutOfRangeException(nameof(createdAfter));
+
+			var createdOrders = _eaOrderController.GetOrders(new Filter("CreatedDateUtc", createdAfter, FilterCondition.GreaterThan));
```
Comment (melissakendall) [inline Mappers/OrderMapper.cs]: this should also be a constant in app.config

File: Mappers/OrderMapper.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+			}
+		}
+
+		/// <summary>
+		/// Sets the shipping and billing information on a Magento cart.
+		/// Shipping address used is based on the EA location unless the data needed is not available. In this case, data is pulled from the customer.
+		/// Shipping method used will be the one set in App.config. If that shipping method cannot be used for the cart specified, an exception will occur.
+		/// </summary>
+		/// <param name="cartId">Cart to set information on.</param>
+		/// <param name="magentoRegion"></param>
```
Comment (melissakendall) [inline Mappers/OrderMapper.cs]: where do these come from :p redundant comments seem redundant but can be useful

File: Mappers/OrderMapper.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+		{
+			var paymentMethod = new CartAddPaymentMethodResource(cartId, ConfigReader.MagentoPaymentMethod);
+			if (_magentoCartController.GetPaymenMethods(cartId).Any(x => x.code == ConfigReader.MagentoPaymentMethod))
+			{
+				_magentoCartController.AddPaymentMethod(cartId, paymentMethod);
+				return _magentoCartController.CreateOrder(cartId, paymentMethod);
+			}
+			else
+			{
+				throw new Exception(string.Format("Unable to create Order for cart {0}. No payment method mtching {1} found for cart.", cartId, ConfigReader.MagentoPaymentMethod));
```
Comment (melissakendall) [inline Mappers/OrderMapper.cs]: matching* also add a note saying what they might do to solve - in this case "ensure the payment method in the app.config is valid and exists in magento"

File: Models/Magento/Cart/BundleOptionResource.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+
+//// ReSharper disable InconsistentNaming
+namespace MagentoConnect.Models.Magento.Cart
+{
+	[Serializable]
+	public class BundleOptionResource
+	{
+		public int option_id { get; set; }// Bundle option id. ,
```
Comment (melissakendall) [inline Models/Magento/Cart/BundleOptionResource.cs]: not all of these models need comments, but lets be consistent and use the <summary> style

File: Utilities/UrlFormatter.cs [hunk 13]
```diff
@@ -377,5 +498,80 @@ public string EndlessAisleCreatePricingUrl()
...
+		/// <param name="url"></param>
+		/// <param name="filters"></param>
+		/// <returns></returns>
+		public string HypermediaFilterUrl(string url, params Filter[] filters)
+		{
+			if (filters.Length == 0)
+				return url;
+
+			StringBuilder result = new StringBuilder(url);
+			result.Append("?$filter=");
```
Comment (melissakendall) [inline Utilities/UrlFormatter.cs]: constant~

File: Utilities/UrlFormatter.cs [hunk 14]
```diff
@@ -377,5 +498,80 @@ public string EndlessAisleCreatePricingUrl()
...
+
+			var filterList = filters.ToList();
+
+			result.Append(filterList.First());
+			filterList.RemoveAt(0);
+
+			foreach (var filter in filterList)
+			{
+				result.Append(string.Format(" and {0}", filter));
```
Comment (melissakendall) [inline Utilities/UrlFormatter.cs]: "and" is a constant in this situation (it could possibly differ depending on language)

File: Utilities/UrlFormatter.cs [hunk 15]
```diff
@@ -377,5 +498,80 @@ public string EndlessAisleCreatePricingUrl()
...
+			if (filters.Length == 0)
+				return url;
+
+			StringBuilder result = new StringBuilder(url);
+			result.Append("?$filter=");
+
+			var filterList = filters.ToList();
+
+			result.Append(filterList.First());
+			filterList.RemoveAt(0);
```
Comment (melissakendall) [inline Utilities/UrlFormatter.cs]: why remove the first? if there is a legit reason, just add a comment

File: Utilities/UrlFormatter.cs [hunk 16]
```diff
@@ -1,4 +1,7 @@
 ﻿using System;
+using System.Collections.Generic;
```
Comment (melissakendall) [inline Utilities/UrlFormatter.cs]: I will make a tech debt ticket for this, but if we are using #region it points to needing to refactor this, to have a base URL class and an extending class for Magento URLs and EA URLs

File: Utilities/RegexPatterns.cs [hunk 17]
```diff
@@ -1,11 +1,12 @@
...
 namespace MagentoConnect.Utilities
 {
-	class RegexPatterns
+	public class RegexPatterns
 	{
-		public const string SlugPattern = "^[M]\\d+(-[V]\\d+)?(-[E]\\d+)?(-[R]([A-Z]{2}|[A-Z]{4}))?$";
+		//Valid slug formats in Endless Aisle
+		public const string SlugPattern = "[M]\\d+(-[V]\\d+)?(-[E]\\d+)?(-[R]([A-Z]{2}|[A-Z]{4}))?";
+
+		//DateTime.ToString() format for "en-US" culture
```
Comment (melissakendall) [inline Utilities/RegexPatterns.cs]: I would use ```<summary>``` for consistency

File: Utilities/LogUtility.cs [hunk 18]
```diff
@@ -1,35 +1,37 @@
...
 		/// <param name="message">Message to write</param>
 		/// <param name="logType">Type of log to write to</param>
-		public static void Write(string message, Log logType)
+		public static void Write(Log logType, string message)
 		{
 			try
 			{
 				using (StreamWriter streamWriter = File.AppendText(Enums.GetPath(logType)))
 				{
 					streamWriter.WriteLine("-------------------------------");
```
Comment (melissakendall) [inline Utilities/LogUtility.cs]: lets define this as a constant at the top of the page, so the user can change how the file is divided if they choose

File: Utilities/LogUtility.cs [hunk 19]
```diff
@@ -1,35 +1,37 @@
...
-					streamWriter.WriteLine("{0}| {1}", DateTime.Now, message);
+					streamWriter.WriteLine("{0}| {1}", DateTime.Now.ToString(CultureInfo.CreateSpecificCulture("en-US")), message);
```
Comment (melissakendall) [inline Utilities/LogUtility.cs]: this should be an app.config constant for all instances, we will only support en-US for now, but if they want to change it we can make it at least slightly easier. Also, add limitation to README saying only en-US is supported

File: Utilities/Filter.cs [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+		{
+			string valueString;
+			if (Value is int)
+			{
+				valueString = Value.ToString();
+			}
+			else if (Value is DateTime)
+			{
+				//Trim the " at the start and end of the serialized object
+				valueString = string.Format("datetime\'{0}\'", JsonConvert.SerializeObject(Value).Substring(1, JsonConvert.SerializeObject(Value).Length - 2));
```
Comment (melissakendall) [inline Utilities/Filter.cs]: datetime = constant at top of file

File: Utilities/Enums.cs [hunk 21]
```diff
@@ -13,12 +14,52 @@ namespace MagentoConnect.Utilities
...
+		[PathValue("\\orderSyncLog.txt")]
+		OrderSync,
 		[PathValue("\\errorLog.txt")]
 		Error
 	}
 
+	public enum OrderState
+	{
+		/// <summary>
+		/// New Order created.
```
Comment (melissakendall) [inline Utilities/Enums.cs]: +1 excellent

**Comment by melissakendall** [comment]
right after we merge this, I'll update the app.config with the values so it doesn't break =P

File: App.cs [hunk 22]
```diff
@@ -35,53 +39,146 @@ public static void Main()
...
+			if (productsSynced)
+			{
+				Console.WriteLine("Products successfully synced");
+				doOrderSync = true;
+			}
+			else
+			{
+				Console.WriteLine("An error occurred while syncing products to Endless Aisle. Check errorLog.txt for more details.");
+				Console.WriteLine("Continue on to synchronizing Orders to Magento?");
+				doOrderSync = Console.ReadKey().ToString().Equals("y", StringComparison.OrdinalIgnoreCase);
```
Comment (melissakendall) [inline App.cs]: y = constant at top of file

File: App.cs [hunk 23]
```diff
@@ -35,53 +39,146 @@ public static void Main()
...
 				{
-					productsToUpdate = _productMapper.GetMagentoProductsUpdatedAfter(DateTime.Now.AddHours(-1));
+					foreach (var order in ordersToCreate)
+					{
+						lastSync = order.CreatedDateUtc > lastSync ? order.CreatedDateUtc : lastSync;
+						int cartId = _orderMapper.CreateCustomerCart();
+						_orderMapper.AddOrderItemsToCart(order.Id.ToString(), cartId);
+						_orderMapper.SetShippingAndBillingInformationForCart(cartId, _entityMapper.MagentoRegion, _entityMapper.EaLocation, _customerMapper.MagentoCustomer);
+						int orderCreatedId = _orderMapper.CreateOrderForCart(cartId);
+						Console.WriteLine(string.Format("Order with ID {0} in Magento has been created from order {1} in Endless Aisle.", orderCreatedId, order.Id));
```
Comment (melissakendall) [inline App.cs]: I think console. write line uses string.format, so you can probably remove string.format here

File: App.cs [hunk 24]
```diff
@@ -35,53 +39,146 @@ public static void Main()
...
+		/// <summary>
+		/// Performs the sync for Magento products orders to Endless Aisle.
+		/// </summary>
+		/// <returns>If the sync was susscessful</returns>
+		private static bool ProductSync()
+		{
+			try
+			{
+				DateTime lastSync = GetTimeForSync(Log.ProductSync);
+				IEnumerable<ProductResource> productsToUpdate = _productMapper.GetMagentoProductsUpdatedAfter(lastSync);
```
Comment (melissakendall) [inline App.cs]: suggestion:
List<ProductResource> productsToUpdate = (List<ProductResource>)

because looping an IEnumerable can cause difficulties

File: Tests/Tests/Controllers/Magento/RegionControllerTests.cs [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+
+		//Tests go here
+
+		/// <summary>
+		/// This test ensures the request does not error.
+		/// </summary>
+		[TestMethod]
+		public void RegionController_GetCountries()
+		{
+			IEnumerable<CountryResource> countries = _regionController.GetCountries();
```
Comment (melissakendall) [inline Tests/Tests/Controllers/Magento/RegionControllerTests.cs]: Assert.IsNotNull()

File: Tests/Tests/Controllers/Magento/CustomerControllerTests.cs [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+
+		//Tests go here
+
+		/// <summary>
+		/// This test ensures the request does not error. If this test fails, either the CustomerId provided is invalid or the request errored 
+		/// </summary>
+		[TestMethod]
+		public void CustomerController_GetCustomer()
+		{
+			CustomerResource customer = _customerController.GetCustomer(CustomerId);
```
Comment (melissakendall) [inline Tests/Tests/Controllers/Magento/CustomerControllerTests.cs]: Assert things~ Also, I realize this is against TDD, but I would add two checks here - a test for CustomerId valid and a test that the request was successful

**Comment by melissakendall** [comment]
Add "Assert" statements to your tests, assert when things should be null, empty, not null, etc

File: Tests/Tests/Controllers/EndlessAisle/OrderControllerTests.cs [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Tests.Controllers.EndlessAisle
+{
+	[TestClass]
+	public class OrderControllerTests
+	{
+		//Private variables go here
+		private OrderController _orderController;
+		private const string OrderId = "cdd26b8f-4ed1-409d-9984-982e081c425e";
```
Comment (melissakendall) [inline Tests/Tests/Controllers/EndlessAisle/OrderControllerTests.cs]: I would say this is not a valid test - if I ran this, it wouldn't work, right? Because I don't have this specific order. So instead, I would create an Order here, or at least "get" an Order you know is valid, and then perform the test

File: Controllers/Magento/ProductController.cs (hunk 1 above)
Comment (seanbocz) [inline Controllers/Magento/ProductController.cs]: Inventory was used because of the documentation on the Magento APIs
